source/extensions/transport_sockets/tls/context_impl.cc

```diff
@@ -595,6 +601,33 @@ std::vector<Ssl::PrivateKeyMethodProviderSharedPtr> ContextImpl::getPrivateKeyMe
   return providers;
 }
 
+void ContextImpl::updateCertStats(Stats::Store& store) {
+  // Add CA cert expiration time.
+  std::vector<Ssl::CertificateNameExpiration> certs_expiration_time =
+      cert_validator_->getExpirationOfCerts();
+
+  // Add TLS certs' expiration time.
+  for (const auto& ctx : tls_contexts_) {
```

Comment:
Instead of discovering all the certs here, can you create the stats for them when the cert is added (and delete the stat when the cert is deleted)?

With this approach, you'll need to do a diff and figure out which ones have been deleted and remove them from the map. And this will all have to be computed every time the timer fires.

Reply:
Yea sounds good. I moved creation of certStat for TLS certs into the TlsContext struct and creation of certStat for CA certs into the cert validator which should delete the stat when the cert is deleted. 
Though I don't expect CA certs to be deleted(?)